cpp/dolfin/fem/FormIntegrals.cpp

```diff
@@ -94,13 +172,151 @@ int FormIntegrals::num_integrals(FormIntegrals::Type type) const
+const std::vector<std::int32_t>&
+FormIntegrals::integral_domains(FormIntegrals::Type type, unsigned int i) const
+{
+  switch (type)
+  {
+  case Type::cell:
+    if (i >= _cell_integral_domains.size())
+      throw std::runtime_error("Invalid cell integral:" + std::to_string(i));
+    return _cell_integral_domains[i];
+  case Type::exterior_facet:
+    return _exterior_facet_integral_domains[i];
+  case Type::interior_facet:
+    return _interior_facet_integral_domains[i];
+  default:
+    throw std::runtime_error("FormIntegral type not supported.");
+  }
+
+  return _cell_integral_domains[i];
+}
+//-----------------------------------------------------------------------------
+void FormIntegrals::set_domains(FormIntegrals::Type type,
+                                const mesh::MeshFunction<std::size_t>& dOmega)
+{
+  std::shared_ptr<const mesh::Mesh> mesh = dOmega.mesh();
+
+  if (type == Type::cell)
+  {
+    if (mesh->topology().dim() != dOmega.dim())
+      throw std::runtime_error("Invalid MeshFunction dimension:"
+                               + std::to_string(dOmega.dim()));
+
+    if (_cell_integral_ids.size() == 0)
+      throw std::runtime_error("No cell integrals");
+
+    // Create a reverse map
+    std::map<int, int> cell_id_to_integral;
+    for (unsigned int i = 0; i < _cell_integral_ids.size(); ++i)
+    {
+      if (_cell_integral_ids[i] != -1)
+      {
+        _cell_integral_domains[i].clear();
+        cell_id_to_integral[_cell_integral_ids[i]] = i;
+      }
+    }
+
+    for (unsigned int i = 0; i < dOmega.size(); ++i)
+    {
+      auto it = cell_id_to_integral.find(dOmega[i]);
+      if (it != cell_id_to_integral.end())
+        _cell_integral_domains[it->second].push_back(i);
+    }
+  }
+  else if (type == Type::exterior_facet)
+  {
+    if (mesh->topology().dim() - 1 != dOmega.dim())
+      throw std::runtime_error("Invalid MeshFunction dimension:"
+                               + std::to_string(dOmega.dim()));
+    if (_exterior_facet_integral_ids.size() == 0)
+      throw std::runtime_error("No exterior facet integrals");
+
+    // Create a reverse map
+    std::map<int, int> facet_id_to_integral;
+    for (unsigned int i = 0; i < _exterior_facet_integral_ids.size(); ++i)
+    {
+      if (_exterior_facet_integral_ids[i] != -1)
+      {
+        _exterior_facet_integral_domains[i].clear();
+        facet_id_to_integral[_exterior_facet_integral_ids[i]] = i;
+      }
+    }
+
+    for (unsigned int i = 0; i < dOmega.size(); ++i)
+    {
+      auto it = facet_id_to_integral.find(dOmega[i]);
+      if (it != facet_id_to_integral.end())
+        _exterior_facet_integral_domains[it->second].push_back(i);
+    }
+  }
+  else
+  {
+    throw std::runtime_error("FormIntegral type not supported.");
+  }
+}
+//-----------------------------------------------------------------------------
+void FormIntegrals::set_default_domains_from_mesh(
+    std::shared_ptr<const mesh::Mesh> mesh)
```

Comment:
Looks like `mesh` should be passed as a const reference rather than a shared_ptr.